src/core/files.ml

```diff
@@ -0,0 +1,280 @@
+    val iter : (Path.t -> file_path -> unit) -> t -> unit
+  end =
+  struct
+    type t = Node of file_path option * t StrMap.t
+
+    let empty = Node(None, StrMap.empty)
+
+    exception Already_mapped
+
+    let set_root path (Node(po, m)) =
+      match po with
+      | None    -> Node(Some(path), m)
+      | Some(_) -> raise Already_mapped
+
+    let rec singleton ks path =
+      match ks with
+      | []      -> Node(Some(path), StrMap.empty)
+      | k :: ks -> Node(None, StrMap.singleton k (singleton ks path))
+
+    let rec add ks path (Node(po, m)) =
+      match (po, ks) with
+      | (Some(_), []     ) -> raise Already_mapped
+      | (None   , []     ) -> Node(Some(path), m)
+      | (_      , k :: ks) ->
+          try Node(po, StrMap.add k (add ks path (StrMap.find k m)) m)
+          with Not_found -> Node(po, StrMap.add k (singleton ks path) m)
+
+    exception Root_not_set
+
+    let get ks (Node(po, m)) =
+      let rec get (root, old_ks) ks m =
+        match ks with
+        | []      ->
+            List.fold_left Filename.concat root old_ks
+        | k :: ks ->
+            match StrMap.find k m with
+            | Node(Some(root), m) -> get (root, ks) ks m
+            | Node(None      , m) -> get (root, old_ks) ks m
+            | exception Not_found ->
+                List.fold_left Filename.concat root old_ks
+      in
+      match po with
+      | None       -> raise Root_not_set
+      | Some(root) -> get (root, ks) ks m
+
+    let iter fn m =
+      let rec iter path (Node(po, m)) =
+        Option.iter (fun file -> fn path file) po;
+        StrMap.iter (fun k m -> iter (path @ [k]) m) m
+      in
+      iter [] m
+  end
+
+(** [lib_root] stores the result of the ["--lib-root"] flag when given. *)
+let lib_root : string option Pervasives.ref =
+  Pervasives.ref None
+
+(** [set_lib_root path] sets the library root to the given file [path]. If the
+    given [path] does not refer to a valid (existing) directory the program is
+    terminated with a graceful error message. *)
+let set_lib_root : string -> unit = fun path ->
+  try
+    let path = Filename.realpath path in
+    if not (Sys.is_directory path) then
+      exit_with "Invalid library root ([%s] is not a directory)." path;
+    match Pervasives.(!lib_root) with
+    | None    -> Pervasives.(lib_root := Some(path))
+    | Some(_) -> exit_with "The library root was already set."
+  with Sys_error(_) | Invalid_argument(_) ->
+    exit_with "Invalid library root (no such file or directory [%s])." path
+
+(** [default_lib_root ()] returns the default library root curently configured
+    for the system. It depends on the current Opam switch (if any), and it may
+    or may not correspond to an existing directory. *)
+let default_lib_root : unit -> string = fun _ ->
+  match run_process "opam var lib" with
+  | Some([l]) -> Filename.concat l "lambdapi/lib_root"
+  | _         -> "/usr/local/lib/lambdapi/lib_root"
```

Comment:
Shouldn't we allow the user to set an environment variable such as `LAMBDAPI_LIB_ROOT`?

Reply:
I would say no, at least for now. There is already a command line option to change the library root, and that is all we need for development mode. Once installed, the choice if library root should always be consistent if we want to be able to have well-integrated tooling.

Of course, we will always be able to reconsider if we want to have several cohabiting (installed) versions of Lambdapi in the same Opam switch. But for now we do not need that.